src/rustup.rs

```diff
@@ -0,0 +1,142 @@
+    /// copy them into `s3://dev-static-rust-lang-org/rustup/dist/`.
+    ///
+    /// When a new release is done and this method is invoked, it downloads the artifacts from that
+    /// bucket (which must always be set as the `DOWNLOAD_BUCKET` variable). A copy of the artifacts
+    /// is archived in `s3://${UPLOAD_BUCKET}/rustup/archive/${version}/`, where `version` is passed
+    /// to this program as a command-line argument. `UPLOAD_BUCKET` can either be the `dev-static`
+    /// or the `static` bucket.
+    ///
+    /// If the release is for the `stable` channel, the artifacts are also copied to the `dist/`
+    /// path in the `UPLOAD_BUCKET` bucket. The `dist/` path is used by the `rustup` installer to
+    /// download the latest release.
+    ///
+    /// Then, the `release-stable.toml` manifest is updated with the new version and copied to
+    /// `s3://${UPLOAD_BUCKET}/rustup/release-stable.toml`.
+    ///
+    /// [release process]: https://rust-lang.github.io/rustup/dev-guide/release-process.html
+    /// [rust-lang/rustup]: https://github.com/rust-lang/rustup
+    pub fn promote_rustup(&mut self) -> anyhow::Result<()> {
+        println!("Checking channel...");
+        if self.config.channel != Channel::Stable && self.config.channel != Channel::Beta {
+            return Err(anyhow!(
+                "promoting rustup is only supported for the stable and beta channels"
+            ));
+        }
+
+        // Download the rustup artifacts from S3
+        println!("Downloading artifacts from dev-static...");
+        let dist_dir = self.download_rustup_artifacts()?;
+
+        // Archive the artifacts
+        println!("Archiving artifacts...");
+        self.archive_rustup_artifacts(&dist_dir)?;
+
+        if self.config.channel == Channel::Stable {
+            // Promote the artifacts to the release bucket
+            println!("Promoting artifacts to dist/...");
+            self.promote_rustup_artifacts(&dist_dir)?;
+        }
+
+        // Update the release number
+        println!("Updating version and manifest...");
+        self.update_rustup_release()?;
+
+        Ok(())
+    }
+
+    fn download_rustup_artifacts(&mut self) -> Result<PathBuf, Error> {
+        let dl = self.dl_dir().join("dist");
+        // Remove the directory if it exists, otherwise just ignore.
+        let _ = fs::remove_dir_all(&dl);
+        fs::create_dir_all(&dl)?;
+
+        run(self
+            .aws_s3()
+            .arg("cp")
+            .arg("--recursive")
+            .arg("--only-show-errors")
+            .arg(&self.s3_artifacts_url("dist/"))
+            .arg(format!("{}/", dl.display())))?;
+
+        Ok(dl)
+    }
+
+    fn archive_rustup_artifacts(&mut self, dist_dir: &Path) -> Result<(), Error> {
+        let version = self
+            .current_version
+            .as_ref()
+            .ok_or_else(|| anyhow!("failed to get current version for rustup release"))?;
```

Comment:
This doesn't work, since `self.current_version` is `None` when this runs. Apparently, the version is set much later in the normal release process.

The easiest workaround might be setting a new `PROMOTE_RELEASE_RUSTUP_VERSION` environment variable, but that would need to be updated manually before running the CodeBuild project. That feels risky, since it's very easy to forget this step and right now we would just override existing artifacts. 😬

Reply:
I think we should figure out how to persist the intended version with the artifacts. That should be possible similarly to how rust has a version containing file checked in, and we can upload that directly to the S3 bucket in rustup's existing "CI".

Reply:
I agree with Mark that the best approach here is for rustup to store the version number in a file in the repository, so that they can update it on their own.

In general, regarding arguments, I don't see any problem with sticking with environment variables. We never invoke the CodeBuild job directly, we always go through https://github.com/rust-lang/simpleinfra/blob/master/release-scripts/promote-release.py, which has a CLI parser and then sets the environment variables.

Reply:
An early step in our release process is a version bump in `Cargo.toml`(e.g. https://github.com/rust-lang/rustup/pull/3773/commits/cfca13cc0bb37ed936dc4ff69086a68ad148bfa2). Is that useful for this particular purpose?

Reply:
The version is now read from the `Cargo.toml` in the latest commit on the `stable` branch.

Reply:
@djc Now that we're at it, maybe it makes sense to use workspace-wide version numbers? I don't see how `rustup-download` can be at a different version from `rustup` itself.

Reply:
@jdno FYI: https://github.com/rust-lang/rustup/pull/4041 now uses a workspace-wide version number.